hoodie-client/src/main/java/com/uber/hoodie/table/HoodieMergeOnReadTable.java

```diff
@@ -74,21 +78,86 @@
+  /**
+   * UpsertPartitioner for MergeOnRead table type, this allows auto correction of small parquet
+   * files to larger ones without the need for an index in the logFile.
+   */
+  class MergeOnReadUpsertPartitioner extends HoodieCopyOnWriteTable.UpsertPartitioner {
+
+    MergeOnReadUpsertPartitioner(WorkloadProfile profile) {
+      super(profile);
+    }
+
+    @Override
+    protected List<SmallFile> getSmallFiles(String partitionPath) {
+
+      List<SmallFile> smallFileLocations = new ArrayList<>();
+      // Init here since this class (and member variables) might not have been initialized
+      HoodieTimeline commitTimeline = getCompletedCommitTimeline();
+
+      if (!commitTimeline.empty()) {
+        HoodieInstant latestCommitTime = commitTimeline.lastInstant().get();
+        // find smallest file in partition and append to it
+        Optional<FileSlice> smallFileSlice = getRTFileSystemView()
+            .getLatestFileSlicesBeforeOrOn(partitionPath, latestCommitTime.getTimestamp())
+            .filter(fileSlice -> fileSlice.getLogFiles().count() < 1 &&
+                fileSlice.getDataFile().get().getFileSize() < config.getParquetSmallFileLimit())
+            .sorted((FileSlice left, FileSlice right) ->
+                left.getDataFile().get().getFileSize() < right.getDataFile().get().getFileSize() ? -1 : 1)
+            .findFirst();
+
+        if(smallFileSlice.isPresent()) {
+            String filename = smallFileSlice.get().getDataFile().get().getFileName();
+            SmallFile sf = new SmallFile();
+            sf.location = new HoodieRecordLocation(FSUtils.getCommitTime(filename),
+                FSUtils.getFileId(filename));
+            sf.sizeBytes = smallFileSlice.get().getDataFile().get().getFileSize();
+            smallFileLocations.add(sf);
+            smallFileIds.add(FSUtils.getFileId(filename));
+          }
+        }
+
+      return smallFileLocations;
+    }
+
+    public List<String> getSmallFileIds() {
+      return smallFileIds;
+    }
+  }
+
+  @Override
+  public Partitioner getUpsertPartitioner(WorkloadProfile profile) {
+    if (profile == null) {
+      throw new HoodieUpsertException("Need workload profile to construct the upsert partitioner.");
+    }
+    mergeOnReadUpsertPartitioner = new MergeOnReadUpsertPartitioner(profile);
+    return mergeOnReadUpsertPartitioner;
+  }
+
   @Override
   public Iterator<List<WriteStatus>> handleUpdate(String commitTime, String fileId,
       Iterator<HoodieRecord<T>> recordItr) throws IOException {
     logger.info("Merging updates for commit " + commitTime + " for file " + fileId);
-    HoodieAppendHandle<T> appendHandle =
-        new HoodieAppendHandle<>(config, commitTime, this, fileId, recordItr);
-    appendHandle.doAppend();
-    appendHandle.close();
-    return Collections.singletonList(Collections.singletonList(appendHandle.getWriteStatus()))
-        .iterator();
+
+    if(mergeOnReadUpsertPartitioner.getSmallFileIds().contains(fileId)) {
```

Comment:
this is a little unseemly.. but understand why it needs to be done..

Reply:
Yeah, i don't love it either, thought a lot on how to avoid it but couldn't :(